File: Week_10 毕业总结/394/decode_string.go
package main

import (
	"strconv"
	"strings"
)

func decodeString(s string) string {
	stack := make([]string, 0)
	ptr := 0
	for ptr < len(s) {
		cur := s[ptr]
		if cur >= '0' && cur <= '9' {
			digits := getDigits(s, &ptr)
			stack = append(stack, digits)
 		} else if  cur != ']' {
 			stack = append(stack, string(cur))
 			ptr++
		} else {
			ptr++
			sub := make([]string, 0)

			for stack[len(stack)-1] != "[" {
				sub = append(sub, stack[len(stack)-1])
				stack = stack[:len(stack)-1]
			}
			// reverse
			for i := 0; i < len(sub)/2;i++ {
				sub[i], sub[len(sub)-i-1] = sub[len(sub)-i-1], sub[i]
			}
			stack = stack[:len(stack)-1]
			repeatTime, _ := strconv.Atoi(stack[len(stack)-1])
			stack[len(stack)-1] = strings.Repeat(getString(sub), repeatTime)
		}
	}
	return getString(stack)
}

func getDigits(s string, ptr *int) string {
	ret := ""
	for ; s[*ptr] >= '0' && s[*ptr] <= '9'; *ptr++ {
		ret += string(s[*ptr])
	}
	return ret
}

func getString(v []string) string{
	return strings.Join(v, "")
}

File: Week_10 毕业总结/394/decode_string.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        i = 0

        def get_digits(p):
            n = ""
            while '0' <= s[p] <= '9':
                n += s[p]
                p += 1
            return n

        while i < len(s):
            if '0' <= s[i] <= '9':
                digits = get_digits(i)
                i += len(digits)
                stack.append(digits)
            elif s[i] != ']':
                stack.append(s[i])
                i += 1
            else:
                i += 1
                sub = ""
                while stack[-1] != "[":
                    sub = stack.pop() + sub
                stack.pop()
                repeat_time = int(stack.pop())
                stack.append(sub * repeat_time)

        return "".join(stack)
